virttest/utils_net.py

```diff
@@ -3313,9 +3313,14 @@ def parse_arp(session=None, timeout=60.0, **dargs):
                      **dargs).strip().split('\n')
 
     for line in arp_cache:
-        mac = line.split()[3]
-        ip = line.split()[0]
-        flag = line.split()[2]
+        parsed_elements = line.split()
+        try:
+            mac = parsed_elements[3]
+            ip = parsed_elements[0]
+            flag = parsed_elements[2]
+        except IndexError:
+            # Skip possible empty lines
+            continue
```

Comment:
Normally, it looks not possible to get empty lines here unless some bad thing is preventing us for reading the entry file properly, therefore I'd suggest raising an exact error (`RuntimeError` or `OSError`?) instead, what do you think?

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/ipv4/arp.c?id=038101e6b2cd5c55f888f85db42ea2ad3aecb4b6#n1385

Reply:
Yes, you're right. This way we'd end up shadowing the root cause, and raising another exception which wouldn't be directly related to the real issue. I think that a `VMAddressVerificationError` would be the one being raised at the end, and effectively, it doesn't give much information about the reason behind the failure.


I was having a look on different `OSError` and `RuntimeErrors` and I'm not sure if I could use them to give the needed proper information about the issue. Maybe we should be implementing another exception on `virttest.utils_net`? Something that would go like:
``` python
ArpParseError(NetError):

    def __init__(self, remote=False):
        """ :param remote: Indicates which was the arp table
                           that was being parsed
                           (the guest's or the host's)
            :type remote: bool
        """
        self.remote = remote

    def __str__(self):
        return ("Error when parsing %s arp table. It seemed to be empty. "
                "An OS issue should be the reason behind this." % \
                ("remote" if self.remote else "local host"))
``` 
However, I don't think it would give us too much further information about the root cause of the failure, but that it would stop the execution and explain why it couldn't keep running. What do you think?

Reply:
> However, I don't think it would give us too much further information about the root cause of the failure, but that it would stop the execution and explain why it couldn't keep running. What do you think?

I just checked the commit history related to this function and found that at the first time the function was introduced without using `process.getoutput` as its command runner (https://github.com/avocado-framework/avocado-vt/commit/eb004798d6850137d4ebbd9e9cf4f43f177158d9), it just used Python's builtin file IO APIs to access the proc entry directly. At that stage, it would indicate to some critical OS failure which could not be recovered if we could get an `IndexError` from the above lines, and thus I think an `OSError` would be something we want to tell to the end users.

Afterward, https://github.com/avocado-framework/avocado-vt/commit/e1d50763511aa9848c6714adf66546102b6349bb changed the manner for the extension of some remote execution proposal. However, it brought the overhead and vulnerabilities from involving an extra child process (by invoking `process.getoutput`), which also makes us feel hard to give the end users further information about the root cause.

So, first of all I think maybe we can try to replace `process.getoutput` back to Python's builtin file IO APIs, to reduce the overhead. And then we can catch the `IndexError`s here and throw out `OSError`s for pointing that some really bad thing happened and nothing was received from the output but garbage.

Please let me know your opinion, thanks!

Reply:
> At that stage, it would indicate to some critical OS failure which could not be recovered (...) I think an OSError would be something we want to tell to the end users.

I really see your point now. I think that what you said on
> ...involving an extra child process, which also **makes us feel hard to give the end users further information about the root cause**.

is key.

I'm using the python builtin IO API for reading the files in a local host, and I'll use the session for the case in which remote execution is needed.

 Thank you for the review!